crates/sp-domains/src/storage_proof.rs

```diff
@@ -0,0 +1,273 @@
+/// Any change in key or value there should be changed here accordingly.
+pub struct RuntimeRegistryStorage<Number, Hash>(PhantomData<(Number, Hash)>);
+impl<Number, Hash> StorageMap<RuntimeId, RuntimeObject<Number, Hash>>
+    for RuntimeRegistryStorage<Number, Hash>
+where
+    Number: FullCodec + TypeInfo + 'static,
+    Hash: FullCodec + TypeInfo + 'static,
+{
+    type Query = Option<RuntimeObject<Number, Hash>>;
+    type Hasher = Identity;
+
+    fn module_prefix() -> &'static [u8] {
+        "Domains".as_ref()
+    }
+
+    fn storage_prefix() -> &'static [u8] {
+        "RuntimeRegistry".as_ref()
+    }
+
+    fn from_optional_value_to_query(v: Option<RuntimeObject<Number, Hash>>) -> Self::Query {
+        v
+    }
+
+    fn from_query_to_optional_value(v: Self::Query) -> Option<RuntimeObject<Number, Hash>> {
+        v
+    }
+}
+
+// The domain runtime code with storage proof
+//
+// NOTE: usually we should use the parent consensus block hash to `generate` or `verify` the
+// domain runtime code because the domain's `set_code` extrinsic is always the last extrinsic
+// to execute thus the domain runtime code will take effect in the next domain block, in other
+// word the domain runtime code of the parent consensus block is the one used when construting
+// the `ExecutionReceipt`.
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq, TypeInfo)]
+pub struct DomainRuntimeCodeWithProof {
+    /// Storage witness of the `domain_id` -> `runtime_id` in the `DomainRuntimeMap` state
+    pub domain_runtime_storage_proof: StorageProof,
+    /// Storage witness of the `RuntimeObject` in the `RuntimeRegistry` state
+    pub runtime_registry_storage_proof: StorageProof,
+}
+
+impl DomainRuntimeCodeWithProof {
+    /// Generate `DomainRuntimeCodeWithProof`
+    ///
+    /// - `runtime_id`: the `RuntimeId` of the domain runtime
+    /// - `block_hash`: the consensus block hash used to get domain runtime code
+    ///
+    /// NOTE: `block_hash` usually should be the parent hash of `ER::consensus_block_hash`
+    /// see the comment of `DomainRuntimeCodeWithProof` for more detail.
+    #[cfg(feature = "std")]
+    #[allow(clippy::let_and_return)]
+    pub fn generate<Block: BlockT, PP: ProofProvider<Block>>(
+        proof_provider: &PP,
+        domain_id: DomainId,
+        runtime_id: RuntimeId,
+        block_hash: Block::Hash,
+    ) -> Result<Self, sp_blockchain::Error> {
+        let domain_runtime_storage_proof = {
+            let key = DomainRuntimeMapStorage::storage_map_final_key(domain_id);
+            let proof = proof_provider.read_proof(block_hash, &mut [key.as_slice()].into_iter())?;
+            proof
+        };
+        let runtime_registry_storage_proof = {
+            let key =
+                RuntimeRegistryStorage::<NumberFor<Block>, Block::Hash>::storage_map_final_key(
+                    runtime_id,
+                );
+            let proof = proof_provider.read_proof(block_hash, &mut [key.as_slice()].into_iter())?;
```

Comment:
I really dont like this. You are providing a storage proof that carries a full runtime code and as a result storage proof and Fraud proof will be very big. This can be optimized if the you providing a proof for runtime hash or something similar and fetching it from the consensus state instead

Reply:
Well, the intention of introducing storage proof was to remove the requirement of consensus state for the verifier so it only needs to provide state root to verify the fraud proof. If we can fetch the runtime code from consensus state why do we even need the runtime hash? the `bad_receipt.consensus_bloak_hash` can already identify the correct runtime code.

Reply:
Anything works as long as the storage proof does not carry the runtime code as is

Reply:
There are only 2 approaches:
1. The stateless approach requires the verifier to provide just the state root, so the fraud proof needs to contain all the necessary data including the domain runtime code
2. The stateful approach requires the verifier to have all the data necessary for verification, so the fraud proof doesn't need to contain any storage proof/state at all

While this PR takes approach 1, we take approach 2 in general through host function, so I would say it is more about whether we should:
- Leave this PR as it is, and refactor based on approach 2 after the host function is landed
- Or rework this PR to take approach 2 completely

While I'm more prefer option 1 to unblock this PR since it has been open so long, I'm also okay with option 2.

Reply:
I personally prefer going the route of using the host functions since we already made a decision on this. Its just very hard to review changes with storage proofs that will discarded in followup PR immediately.

> While I'm more prefer option 1 to unblock this PR since it has been open so long

This should never be a concern. If a specific PR is open for long time doesn't mean the approach you have taken is wrong instead reviewers are just not clear on the approach or there are other approaches that makes more sense to the reviewer. Having a PR open for long time should not be on you and it is the responsibility of the reviewer and PR author to figure out a way forward.